core/chains/evm/logpoller/log_poller.go

```diff
@@ -0,0 +1,380 @@
+			start = fromBlock
+		case <-tick:
+			tick = time.After(utils.WithJitter(lp.pollPeriod))
+			if start == 0 {
+				lastProcessed, err := lp.orm.SelectLatestBlock(pg.WithParentCtx(lp.ctx))
+				if err != nil {
+					if !errors.Is(err, sql.ErrNoRows) {
+						lp.lggr.Warnw("unable to get starting block", "err", err)
+						continue
+					}
+					// Otherwise this is the first poll _ever_ on a new chain.
+					// Only safe thing to do is to start at finality depth behind tip.
+					latest, err := lp.ec.BlockByNumber(context.Background(), nil)
+					if err != nil {
+						lp.lggr.Warnw("unable to get latest for first poll", "err", err)
+						continue
+					}
+					// Do not support polling chains with don't even have finality depth worth of blocks.
+					// Could conceivably support this but not worth the effort.
+					if int64(latest.NumberU64()) < lp.finalityDepth {
+						lp.lggr.Criticalw("insufficient number of blocks on chain, log poller exiting", "err", err, "latest", latest.NumberU64())
+						return
+					}
+					start = int64(latest.NumberU64()) - lp.finalityDepth + 1
+					continue
+				}
+				start = lastProcessed.BlockNumber + 1
+				continue
+			}
+			start = lp.PollAndSaveLogs(lp.ctx, start)
+		}
+	}
+}
+
+func min(a, b int64) int64 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func convertLogs(chainID *big.Int, logs []types.Log) []Log {
+	var lgs []Log
+	for _, l := range logs {
+		lgs = append(lgs, Log{
+			EvmChainId: utils.NewBig(chainID),
+			LogIndex:   int64(l.Index),
+			BlockHash:  l.BlockHash,
+			// We assume block numbers fit in int64
+			// in many places.
+			BlockNumber: int64(l.BlockNumber),
+			EventSig:    l.Topics[0].Bytes(), // First topic is always event signature.
+			Topics:      convertTopics(l.Topics),
+			Address:     l.Address,
+			TxHash:      l.TxHash,
+			Data:        l.Data,
+		})
+	}
+	return lgs
+}
+
+func convertTopics(topics []common.Hash) [][]byte {
+	var topicsForDB [][]byte
+	for _, t := range topics {
+		topicsForDB = append(topicsForDB, t.Bytes())
+	}
+	return topicsForDB
+}
+
+func (lp *LogPoller) backfill(ctx context.Context, start, end int64) int64 {
+	for i := start; i <= end; i += lp.backfillBatchSize {
+		var (
+			logs []types.Log
+			err  error
+		)
+		from := i
+		to := min(i+lp.backfillBatchSize-1, end)
+		utils.RetryWithBackoff(ctx, func() bool {
+			logs, err = lp.ec.FilterLogs(ctx, ethereum.FilterQuery{
+				FromBlock: big.NewInt(from),
+				ToBlock:   big.NewInt(to),
+				Addresses: lp.filterAddresses(),
+				Topics:    lp.filterTopics(),
+			})
+			if err != nil {
+				lp.lggr.Warnw("Unable query for logs, retrying", "err", err, "from", from, "to", to)
+				return true
+			}
+			return false
+		})
+		if len(logs) == 0 {
+			continue
+		}
+		lp.lggr.Infow("Backfill found logs", "from", from, "to", to, "logs", len(logs))
+		utils.RetryWithBackoff(ctx, func() bool {
+			if err := lp.orm.InsertLogs(convertLogs(lp.ec.ChainID(), logs)); err != nil {
```

Comment:
If we have a db issue, this could get stuck indefinitely.
I like the idea of reporting not Healthy via Healthy() in such cases.

But ok if you want to just add a comment, and can do it later if we feel this becomes a more common problem.